rfd/0088-passwordless-windows.md

```diff
@@ -0,0 +1,216 @@
+way.
+
+## Details
+
+Windows Hello and FIDO devices support is implemented via win32 API which is
+installed on Windows 10 19H1 or later.
+More details about API can be found in [API spec](https://github.com/microsoft/webauthn)
+or [documentation][webauthn API windows docs].
+
+```mermaid
+flowchart TB
+  WRP[WebAuthn RP]
+  subgraph CD[Client Device]
+    BC[Browser Client]
+    NC[Native Client]
+    subgraph WADLL[WEBAUTHN.dll]
+        PA[Platorm Authenticator]
+        CPA[Cross-platform Authenticator]
+    end
+  end
+  subgraph RA[Roaming Authenticator]
+    CTAP2[CTAP2 API]
+  end
+
+
+  WRP --> CD
+  BC --> WADLL
+  NC --> WADLL
+
+  CPA -. USB/NFC/BLE .-> CTAP2
+```
+
+`Webauthn.dll` is designed to work with Webauthn protocol, that's why
+integration with it results moslty in passing parameters from and back to
+server.
+
+<!--
+TODO(tobiaszheller): Check if RS256 requires any code changes on server side.
+-->
+Windows Hello works only with RS256 algorithm, while both FIDO2 and TouchID uses
+ES256.
+
+**Important:** Windows system without `webauthn.dll` (older than Windows 10 19H1)
+won't be supported.
+
+If `webauthn.dll` is not found, `tsh` should disable Windows Hello or FIDO
+devices support.
+Microsoft provides 4 versions of webauthn API and all should be supported.
+
+Interactions (touch messages and pin inputs) with Windows Hello or FIDO devices
+is done via Windows UI.
+
+Windows Hello works with both Microsoft online and offline accounts.
+
+### Registration
+
+The proposed UX is similar to the current experience:
+
+```shell
+$ tsh mfa add
+> Choose device type [TOTP, WEBAUTHN]: webauthn
+> Enter device name: win-hello
+> Tap any *registered* security key or enter a code from a *registered* OTP device: <taps>
+<system shows webauthn prompt>
+> MFA device "win-hello" added.
+```
+
+The way to specify if windows hello or FIDO device should be used is via
+mfa-mode: `--mfa-mode={auto,platform,cross-platform}`. Auto defaults to platform
+in Windows Hello.
+
+`tsh` performs the key registration process, setting all parameters for a
+passwordless / resident key via `WebAuthNAuthenticatorMakeCredential` from
+`webauthn.dll`.
+
+<!--
+TODO(codingllama): should we add here some code samples? same for authenticaiton
+-->
+
+In case of a registration failure, `tsh` must do its best to delete the
+created-but-not-registered credential. Credential removal is only possible
+using webauthn API version 4+.
+
+### Authentication
+
+Authentication depends on server settings (otp, webauthn, passwordless).
+
+`tsh` performs the authentication process, setting all parameters for a
+passwordless / resident key via `WebAuthNAuthenticatorGetAssertion` from
+`webauthn.dll`.
+
+The way to specify if Windows Hello or FIDO device should be used is via
+mfa-mode: `tsh --mfa-mode={auto,platform,cross-platform}` - choose whether to
+use platform or cross-platform MFA. Auto defaults to platform in Windows Hello.
```

Comment:
unfortunately no, but if you pass auto and only cross-platform was specified, it will use cross-platform.
 (at least in API 3, in API 4 they added method for managing credentials but it is not yet released in windows 11).

Reply:
Ack.